### PR TITLE
Idempotence assumption

### PR DESCRIPTION
**Pull request by couteiral**
#### Brief description of what is fixed or changed
Implemented `idempotent` assumption.

#### Other comments
When a symbol `x` is assumed to be `idempotent`, any non-zero power `x^k` will evaluate to x.

#### Release Notes

* assumptions
  * implemented `idempotency` assumption
* matrices
  * implemented assumptions in `MatrixSymbol`

### CODE REVIEW

**Comment by asmeurer** [comment]
This needs tests.

File: sympy/core/power.py [hunk 1]
```diff
@@ -263,6 +263,11 @@ def __new__(cls, b, e, evaluate=None):
                     b = -b
                 elif e.is_odd:
                     return -Pow(-b, e)
+            if b.is_idempotent:
+                if e is S.Zero:
```
Comment (asmeurer) [inline sympy/core/power.py]: The exponent needs to be positive. If `x*x = x`, and `x*y = 1`, then `x*x*y = x*1 = x*y = x = 1`, so idempotents are only invertible if they are the identity.

For example, this matrix example from [Wikipedia](https://en.wikipedia.org/wiki/Idempotent_matrix), `Matrix([[2, -2, -4], [-1, 3, 4], [1, -2, -3]])` is not invertible.

**Comment by asmeurer** [comment]
A concern with this is that idempotent usually refers to matrices, but the way the (new) assumption is set up I'm not sure if it will apply to a matrix (maybe it can work, I'll have to think about it).

File: sympy/assumptions/ask.py [hunk 2]
```diff
@@ -68,6 +68,22 @@ def antihermitian(self):
         # TODO: Add examples
         return Predicate('antihermitian')
 
+    @predicate_memo
+    def idempotent(self):
+        """
+        Idempotent predicate.
+
+        ``Q.idempotent(x)`` is true iff ``x^n=x`` for non-negative ``x``.
```
Comment (oscarbenjamin) [inline sympy/assumptions/ask.py]: Should this be non-negative n rather than x?

File: sympy/core/assumptions.py [hunk 3]
```diff
@@ -106,6 +106,10 @@
         object belongs to the field of hermitian
         (antihermitian) operators.
 
+    idempotent
+        non-zero power of the element is equal
+        to the element
```
Comment (oscarbenjamin) [inline sympy/core/assumptions.py]: Maybe nonzero should be positive.

Comment (couteiral) [reply sympy/core/assumptions.py]: Corrected.

File: sympy/assumptions/ask.py (hunk 2 above)
Comment (couteiral) [reply sympy/assumptions/ask.py]: Indeed, corrected.

File: sympy/core/power.py (hunk 1 above)
Comment (couteiral) [reply sympy/core/power.py]: Corrected

**Comment by couteiral** [comment]
> A concern with this is that idempotent usually refers to matrices, but the way the (new) assumption is set up I'm not sure if it will apply to a matrix (maybe it can work, I'll have to think about it).

This pull request was prompted by a problem that did not involve matrices, so I did not consider that. I will have a look at how to implement it, but I recognise it is probably a different pull request. Let me know what you think.

**Comment by couteiral** [comment]
Could this be merges?

File: sympy/core/tests/test_power.py [hunk 4]
```diff
@@ -452,3 +452,20 @@ def test_better_sqrt():
...
+    assert x**2     == x
+    assert x**125   == x
+    assert x**n     == x
+    assert x**(n+1) == x
+
+    # empty product
+    assert x**0     == 1
+
+    # not-necessarily-positive powers are left untouched
+    assert x**y     != x
```
Comment (oscarbenjamin) [inline sympy/core/tests/test_power.py]: Can you assert what this is equal to e.g. `Pow(x, y, evaluate=False)`?

File: sympy/assumptions/ask.py [hunk 5]
```diff
@@ -68,6 +68,22 @@ def antihermitian(self):
...
+
+        ``Q.idempotent(x)`` is true iff ``x^n=x`` for positive ``n``.
+
+        References
+        ==========
+
+        .. [1] https://en.wikipedia.org/wiki/Idempotence
+
+        """
+        # TODO: Add examples
```
Comment (oscarbenjamin) [inline sympy/assumptions/ask.py]: You've left a TODO here. Should that be done now?

**Comment by oscarbenjamin** [comment]
Can this be used for a MatrixSymbol or a Matrix?

What other examples would there be of idempotence? Anything idempotent would have to satisfy `x**2 = x` so for finite complex `x` the only possibilities would be 0 and 1. SymPy would also allow `oo` and `zoo` to satisfy that equation but is that what you want to be considered idempotent?

File: sympy/core/tests/test_power.py (hunk 4 above)
Comment (couteiral) [reply sympy/core/tests/test_power.py]: Done.

File: sympy/assumptions/ask.py (hunk 5 above)
Comment (couteiral) [reply sympy/assumptions/ask.py]: That's a copy-pasting error from the docs for Hermitian and AntiHermitian just above. I have deleted it.

**Comment by couteiral** [comment]
Aside from boolean variables, this can be used e. g. to handle abstract operator algebras where these are idempotents. 

I can see how this would be adapted for MatrixSymbol, not sure to Matrix. @oscarbenjamin and @asmeurer, would you like me to implement this here or do another pull request?

**Comment by couteiral** [comment]
I have implemented idempotence (and therefore assumptions) in MatrixSymbol. Note that matrix assumptions also handle hermiticity and antihermiticity, but assumes non-commutativity regardless of the input; this could be changed but I don't see many cases where it would be useful.

**Comment by couteiral** [comment]
I have updated the release notes, not sure how to make sympy-bot handle that!

**Comment by couteiral** [comment]
Can this be merged?

**Comment by oscarbenjamin** [comment]
Sorry for the long delay getting back. This has merge conflicts now so it can't be merged until those are resolved...

Otherwise this adds a new basic fact to the assumptions system. It's not clear to me if the new fact is needed but this is the way to add it if it is.

File: sympy/matrices/expressions/tests/test_matpow.py [hunk 6]
```diff
@@ -137,3 +137,21 @@ def test_transpose_power():
...
+
+def test_idempotence():
+
+    F = MatrixSymbol('F', n, n, idempotent=True)
+    t = symbols('t', positive=True)
+
+    # positive powers equal the element
+    assert F**2     == F
+    assert F**125   == F
+    assert F**t     == F
```
Comment (oscarbenjamin) [inline sympy/matrices/expressions/tests/test_matpow.py]: Should `t` be an integer here? How does idempotence work with `F**1.5`?

**Comment by czgdp1807** [comment]
Closing in favour of https://github.com/sympy/sympy/pull/18750

**Pull request closed without merging**